addons/binding/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/hw/HwDiscoveryService.java

```diff
@@ -0,0 +1,75 @@
+package org.openhab.binding.lutron.internal.hw;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.lutron.internal.LutronHandlerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
+ * The Discovery Service for Lutron HomeWorks processors. There is no great way to automatically
+ * discovery modules in the legacy HomeWorks processor (that I know of) so this service simply iterates
+ * through possible addresses and asks for status on that address. If it's a valid module, the processor will return
+ * with the dimmer status and it will be discovered.
+ *
+ * @author Andrew Shilliday
+ *
+ */
+public class HwDiscoveryService extends AbstractDiscoveryService {
+    private Logger logger = LoggerFactory.getLogger(HwDiscoveryService.class);
+
+    private final HwSerialBridgeHandler handler;
+
+    public HwDiscoveryService(HwSerialBridgeHandler handler) {
+        super(LutronHandlerFactory.HW_DISCOVERABLE_DEVICE_TYPES_UIDS, 10);
+        this.handler = handler;
+    }
+
+    @Override
+    protected void startScan() {
+        new Thread(() -> {
+            try {
+                logger.debug("Starting scan for HW Dimmers");
+                for (int m = 1; m <= 8; m++) { // Modules
+                    for (int o = 1; o <= 4; o++) { // Outputs
+                        String address = String.format("[01:01:00:%02d:%02d]", m, o);
+                        handler.sendCommand("RDL, " + address);
+                        Thread.sleep(5);
+                    }
+                }
+            } catch (InterruptedException e) {
+                logger.warn("Scan interrupted");
+            }
+        }).start();
+    }
+
+    /**
+     * Called by the bridge when it receives a status update for a dimmer that is not registered.
+     */
+    public void declareUnknownDimmer(String address) {
+        if (address == null) {
+            logger.info("Discovered HomeWorks dimmer with no address");
+            return;
+        }
+        String addressUid = address.replaceAll("[\\[\\]]", "").replaceAll(":", "-");
+        ThingUID bridgeUID = this.handler.getThing().getUID();
+        ThingUID uid = new ThingUID(HwConstants.THING_TYPE_HWDIMMER, bridgeUID, addressUid);
+
+        Map<String, Object> props = new HashMap<>();
+
+        // TODO put this in a constant
```

Comment:
There is a TODO, what should be done and can it be completed before merging the PR?

Reply:
I removed the TODO on this one and am otherwise skipping.  It's been a while but I can't for the life of me figure out what I was talking about.  I suspect it was for an older version of that code.